modules/adwords_appengine/src/main/java/com/google/api/ads/adwords/jaxws/utils/v201809/ServiceQuery.java

```diff
@@ -0,0 +1,301 @@
+   * change the LIMIT clause.
+   *
+   * <p>This method is meant to be used with {@code hasNext()}. The {@code page} is necessary when
+   * using DataService, as its paging mechanism is different from other services. For details, see
+   * https://developers.google.com/adwords/api/docs/guides/bid-landscapes#paging_through_results.
+   *
+   * @param page the criterion bid landscape page whose total number of landscape points will be
+   *     used to increment the index in the LIMIT clause
+   * @return this service query object
+   */
+  public ServiceQuery nextPage(@Nullable CriterionBidLandscapePage page) {
+    if (page == null) {
+      return this;
+    }
+    return nextPage(getTotalLandscapePointsInPage(page));
+  }
+
+  /**
+   * Advances this object's {@link #startIndex} by the specified {@link #pageSize}.
+   *
+   * @return this service query object
+   */
+  private ServiceQuery nextPage(int pageSize) {
+    startIndex = startIndex + pageSize;
+    return this;
+  }
+
+  @Override
+  public boolean hasNext(@Nullable Page page) {
+    return (page == null) || startIndex + pageSize < page.getTotalNumEntries();
+  }
+
+  /**
+   * Checks if there is still an ad group bid landscape page left to query.
+   *
+   * <p>This method is meant to be used with {@link ServiceQuery#nextPage(AdGroupBidLandscapePage)}.
+   * The {@code page} is necessary when using DataService, as its paging mechanism is different from
+   * other services. For details, see
+   * https://developers.google.com/adwords/api/docs/guides/bid-landscapes#paging_through_results.
+   *
+   * @param page the ad group bid landscape page whose total number of landscape points will be used
+   *     to determine if there is still a page left
+   * @return true if there is still a page left
+   */
+  public boolean hasNext(@Nullable AdGroupBidLandscapePage page) {
+    if (page == null) {
+      return true;
+    }
+    return getTotalLandscapePointsInPage(page) >= pageSize;
+  }
+
+  /**
+   * Checks if there is still an criterion bid landscape page left to query.
+   *
+   * <p>This method is meant to be used with {@link
+   * ServiceQuery#nextPage(CriterionBidLandscapePage)}. The {@code page} is necessary when using
+   * DataService, as its paging mechanism is different from other services. For details, see
+   * https://developers.google.com/adwords/api/docs/guides/bid-landscapes#paging_through_results.
+   *
+   * @param page the criterion bid landscape page whose total number of landscape points will be
+   *     used to determine if there is still a page left
+   * @return true if there is still a page left
+   */
+  public boolean hasNext(@Nullable CriterionBidLandscapePage page) {
+    if (page == null) {
+      return true;
+    }
+    return getTotalLandscapePointsInPage(page) >= pageSize;
+  }
+
+  @Override
+  public String toString() {
+    return String.format("%s LIMIT %s,%s", partialAwqlQuery, startIndex, pageSize);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: startIndex, pageSize. This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 158. Consider formatting this data with &#39;%d&#39; instead.